src/executor/tensorrt_pass.cc

```diff
@@ -0,0 +1,583 @@
+        }
+      }
+      for (Node* nonTRTn : set_nonTRTnodes) {
+        uncomp_map[n].erase(nonTRTn);
+      }
+    }
+    std::unordered_set<Node*> set_unused;
+    set_unused.reserve(set_TRTnodes.size());
+
+    for (auto& n : set_TRTnodes) {
+      if (n->nnvmptr->attrs.op != nullptr || params_map->count(n->nnvmptr->attrs.name)) {
+        set_unused.insert(n);
+      }
+    }
+    std::unordered_set<Node*> visited;
+    std::deque<Node*> stack(outputs.begin(), outputs.end());
+    while (!stack.empty()) {
+      Node* vertex = stack.front();
+      stack.pop_front();
+      if (!visited.count(vertex)) {
+        visited.insert(vertex);
+        if (set_unused.count(vertex)) {
+          subgraphs.emplace_back(naive_grow_subgraph(vertex, &set_unused, &uncomp_map));
+        }
+        for (Node* input : vertex->inputs) {
+          stack.emplace_back(input);
+        }
+      }
+    }
+
+    return subgraphs;
+  }
+
+
+ private:
+  friend class Graph;
+
+  bool IsTRTCompatible(nnvm::Node* nodeptr) {
+    if (nodeptr->op() != nullptr) {
+      const std::string op_name = nodeptr->op()->name;
+      if (op_name == "Pooling") {
+        return (nodeptr->attrs.dict.at("pool_type") == "avg" ||
+          nodeptr->attrs.dict.at("pool_type") == "max");
+      } else if (unconditionalTRTop.count(op_name)) {
+        return true;
+      } else if (op_name == "Activation") {
+        return nodeptr->attrs.dict.at("act_type") == "relu" ||
+          nodeptr->attrs.dict.at("act_type") == "tanh" ||
+          nodeptr->attrs.dict.at("act_type") == "sigmoid";
+      }
+      return false;
+    }
+    return true;
+  }
+};  // class BidirectionalGraph
+
+/*!
+ * \brief function which transform std::vector<dmlc::any> back to Attrs (dmlc::any)
+ */
+const std::unordered_set<std::string> BidirectionalGraph::unconditionalTRTop = {
+  "Convolution",
+  "BatchNorm",
+  "elemwise_add",
+  "elemwise_sub",
+  "elemwise_mul",
+  "rsqrt",
+  "pad",
+  "Pad",
+  "mean",
+  "FullyConnected",
+  "Flatten",
+  "SoftmaxOutput",
+};
+
+
+using NodeEntrySet = std::unordered_set<nnvm::NodeEntry, nnvm::NodeEntryHash,
+                                        nnvm::NodeEntryEqual>;
+
+/*!
+ * \brief get the output nodes of the subgraph in the main graph
+ * \return a vector of the output nodes
+*/
+std::vector<nnvm::NodeEntry> GetSubgraphOutputs(Graph g,
+    std::unordered_set<nnvm::Node*> set_subgraph) {
+  std::vector<nnvm::NodeEntry> outputs;
+  NodeEntrySet _outputs;
+  for (auto& e : g.outputs) {
+    if (set_subgraph.count(e.node.get())) {
+      _outputs.insert(e);
+    }
+  }
+  DFSVisit(g.outputs, [&set_subgraph, &_outputs](const nnvm::NodePtr &node){
+    if (!set_subgraph.count(node.get())) {
+      for (auto& e : node->inputs) {
+        if (set_subgraph.count(e.node.get())) {
+          _outputs.insert(e);
+        }
+      }
+    }
+  });
+  outputs.insert(outputs.begin(), _outputs.begin(), _outputs.end());
```

Comment:
A subgraph may have duplicate outputs. Would this making output entries unique cause any problem to the nodes who accept the outputs of the subgraph as inputs? In addition, the entry's order is not kept as in its original graph, and this may affect the order of inputs of the whole graph. Do you think this is a problem here?

Reply:
@mkolod Duplicate outputs of an op are different node entries in the graph (but share the same NDArray in memory). For example, there is a graph, A->B, A->C, final output is B+C. The edges from A to B and C are duplicate entries of A's output. None of them are made using an identity op. Let's say in this case, A is TRT compatible and we make a subgraph out of it. There are two node entries from A, but this algorithm will only give one entry as a result. Is this behavior expected? Have you run any unit tests covering such cases?

Reply:
can you be more specific about A, B, C ? Are they node or node entries (you say that final output is B+C, so are those node entries ?) ? Only Op can / can't be TRT compatible.

Reply:
All of A, B, C, D are operators. Final output entry is the output of B+C.